Applications/Tizen_native/CustomShortcut/src/data.c

```diff
@@ -43,6 +43,200 @@ int parse_route(const char *source, char **route, char **data) {
+
+  data_get_resource_path("mobilenetv2.tflite", model_path, false);
+
+  if (ad->is_pipe_used == true)
+    LOG_W("pipe is currently used, wait until using pipe is done.");
+
+  time_t start = time(NULL);
+  while (ad->is_pipe_used) {
+    if (time(NULL) - start > 10) {
+      LOG_E("pipeline waiting time out");
+      return APP_CONTROL_ERROR_TIMED_OUT;
+    }
+  }
+
+  LOG_D("pipe ready, starting pipeline");
+
+  ad->is_pipe_used = true;
+
+  sprintf(pipe_description,
+          "filesrc location=%s ! pngdec ! "
+          "videoconvert ! videoscale ! "
+          "video/x-raw,width=224,height=224,format=RGB ! "
+          "tensor_converter ! "
+          "tensor_transform mode=arithmetic option=%s ! "
+          "tensor_filter framework=tensorflow-lite model=%s ! "
+          "tensor_sink name=sink",
+          src, "typecast:float32,add:-127.5,div:127.5", model_path);
+
+  LOG_D("setting inference \n pipe: %s", pipe_description);
+  status = ml_pipeline_construct(pipe_description, NULL, NULL, &ad->pipeline);
+
+  if (status != ML_ERROR_NONE) {
+    LOG_E("something wrong constructing pipeline %d", status);
+    ad->is_pipe_used = false;
+    return status;
+  }
+
+  status = ml_pipeline_sink_register(ad->pipeline, "sink", _on_data_receive,
+                                     (void *)ad, &ad->pipe_sink);
+
+  LOG_D("starting inference");
+  status = ml_pipeline_start(ad->pipeline);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("failed to start pipeline %d", status);
+    return status;
+  }
+
+  while (ad->is_pipe_used) {
+    if (time(NULL) - start > 10) {
+      LOG_E("pipeline waiting time out");
+      return APP_CONTROL_ERROR_TIMED_OUT;
+    }
+  }
+
+  LOG_D("stopping pipeline");
+  status = ml_pipeline_stop(ad->pipeline);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("stopping pipeline failed");
+    return status;
+  }
+
+  LOG_D("unregister pipeline");
+  status = ml_pipeline_sink_unregister(ad->pipe_sink);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("unregistering sink failed");
+    return status;
+  }
+  ad->pipe_sink = NULL;
+
+  LOG_D("destroying pipeline");
+  status = ml_pipeline_destroy(ad->pipeline);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("destroying pipeline failed");
+    return status;
+  }
+  ad->pipeline = NULL;
+
+  return status;
+}
+
+int data_extract_feature(appdata_s *ad, const char *dst, bool append) {
+  char png_path[PATH_MAX];
+  cairo_status_t cr_stat = CAIRO_STATUS_SUCCESS;
+  int status = APP_ERROR_NONE;
+
+  data_get_data_path("temp.png", png_path);
+  LOG_D("start writing to png_path: %s ", png_path);
+  cr_stat = cairo_surface_write_to_png(ad->cr_surface, png_path);
+
+  if (cr_stat != CAIRO_STATUS_SUCCESS) {
+    LOG_E("failed to write cairo surface as a file reason: %d", cr_stat);
+    return APP_ERROR_INVALID_PARAMETER;
+  }
+
+  data_get_data_path(dst, ad->pipe_dst);
+
+  LOG_I("start inference to dataset: %s ", ad->pipe_dst);
+  status = _run_nnpipeline(ad, png_path, append);
```

Comment:
This overwrites existing `dataset.dat`. Is this intentional?

Reply:
Yes for now.